nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/EnforceOrder.java

```diff
@@ -0,0 +1,538 @@
+})
+public class EnforceOrder extends AbstractProcessor {
+
+    public static final String ATTR_STARTED_AT = "EnforceOrder.startedAt";
+    public static final String ATTR_EXPECTED_ORDER = "EnforceOrder.expectedOrder";
+    public static final String ATTR_RESULT = "EnforceOrder.result";
+    public static final String ATTR_DETAIL = "EnforceOrder.detail";
+    private static final Function<String, String> STATE_TARGET_ORDER = groupId -> groupId + ".target";
+    private static final String STATE_SUFFIX_UPDATED_AT = ".updatedAt";
+    private static final Function<String, String> STATE_UPDATED_AT = groupId -> groupId + STATE_SUFFIX_UPDATED_AT;
+    private static final Function<String, String> STATE_MAX_ORDER = groupId -> groupId + ".max";
+
+    public static final PropertyDescriptor GROUP_IDENTIFIER = new PropertyDescriptor.Builder()
+        .name("group-id")
+        .displayName("Group Identifier")
+        .description("EnforceOrder is capable of multiple ordering groups." +
+                " 'Group Identifier' is used to determine which group a FlowFile belongs to." +
+                " This property will be evaluated with each incoming FlowFile." +
+                " If evaluated result is empty, the FlowFile will be routed to failure.")
+        .required(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .expressionLanguageSupported(true)
+        .defaultValue("${filename}")
+        .build();
+
+    public static final PropertyDescriptor ORDER_ATTRIBUTE = new PropertyDescriptor.Builder()
+        .name("order-attribute")
+        .displayName("Order Attribute")
+        .description("A name of FlowFile attribute whose value will be used to enforce order of FlowFiles within a group." +
+                " If a FlowFile does not have this attribute, or its value is not an integer, the FlowFile will be routed to failure.")
+        .required(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .expressionLanguageSupported(false)
+        .build();
+
+    public static final PropertyDescriptor INITIAL_ORDER = new PropertyDescriptor.Builder()
+        .name("initial-order")
+        .displayName("Initial Order")
+        .description("When the first FlowFile of a group arrives, initial target order will be computed and stored in the managed state." +
+                " After that, target order will start being tracked by EnforceOrder and stored in the state management store." +
+                " If Expression Language is used but evaluated result was not an integer, then the FlowFile will be routed to failure," +
+                " and initial order will be left unknown until consecutive FlowFiles provide a valid initial order.")
+        .required(true)
+        .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+        .expressionLanguageSupported(true)
+        .defaultValue("1")
```

Comment:
A number of processors (the Split processors, CaptureChange, etc.) use zero as the first index / sequence ID by default, I'm thinking we should default this property to zero as well.